src/threading.c

```diff
@@ -848,6 +848,138 @@ JL_DLLEXPORT int jl_alignment(size_t sz)
     return jl_gc_alignment(sz);
 }
 
+// Heartbeat mechanism for Julia's task scheduler
+// ---
+// Start a thread that does not participate in running Julia's tasks. This
+// thread simply sleeps until the heartbeat mechanism is enabled. When
+// enabled, the heartbeat thread enters a loop in which it waits for the
+// specified heartbeat interval. If, within that interval, `jl_heartbeat()`
+// is *not* called at least once, then the thread calls
+// `jl_print_task_backtraces(0)`.
+
+#ifdef JL_HEARTBEAT_THREAD
+
+#include <time.h>
+
+volatile int heartbeat_enabled;
+uv_sem_t heartbeat_sem;
+int heartbeat_interval_s, report_interval_s, last_report_s;
+_Atomic(int) heartbeats;
+
+JL_DLLEXPORT void jl_print_task_backtraces(int show_done) JL_NOTSAFEPOINT;
+void jl_heartbeat_threadfun(void *arg);
+
+// start the heartbeat thread with heartbeats disabled
+void jl_init_heartbeat(void)
+{
+    uv_thread_t uvtid;
+    heartbeat_enabled = 0;
+    uv_sem_init(&heartbeat_sem, 0);
+    uv_thread_create(&uvtid, jl_heartbeat_threadfun, NULL);
+    uv_thread_detach(&uvtid);
+}
+
+// enable/disable heartbeats
+JL_DLLEXPORT int jl_heartbeat_enable(int heartbeat_s, int report_s)
+{
+    if (heartbeat_s <= 0) {
+        heartbeat_enabled = 0;
+        last_report_s = 0;
+    }
+    else {
+        jl_atomic_store_relaxed(&heartbeats, 0);
+        heartbeat_interval_s = heartbeat_s;
+        report_interval_s = report_s;
+        heartbeat_enabled = 1;
+        uv_sem_post(&heartbeat_sem);
+    }
+    return 0;
+}
+
+// heartbeat
+JL_DLLEXPORT void jl_heartbeat(void)
+{
+    jl_atomic_fetch_add(&heartbeats, 1);
+}
+
+// sleep the thread for the specified interval
+void sleep_for(int secs, int nsecs)
+{
+    struct timespec rqtp, rmtp;
+    rqtp.tv_sec = secs;
+    rqtp.tv_nsec = nsecs;
+    rmtp.tv_sec = 0;
+    rmtp.tv_nsec = 0;
+    for (; ;) {
+        if (nanosleep(&rqtp, &rmtp) == 0)
+            return;
+        rqtp = rmtp;
+    }
+}
+
+uint8_t check_heartbeats(uint8_t gc_state)
+{
+    int hb = jl_atomic_exchange(&heartbeats, 0);
+
+    if (hb <= 0) {
+        if ((jl_hrtime() / 1e9) - last_report_s >= report_interval_s) {
+            jl_task_t *ct = jl_current_task;
+            jl_ptls_t ptls = ct->ptls;
+            jl_gc_safe_leave(ptls, gc_state);
+            jl_print_task_backtraces(0);
+            gc_state = jl_gc_safe_enter(ptls);
+            last_report_s = jl_hrtime() / 1e9;
+        }
+    }
+    else
+        last_report_s = 0;
+
+    return gc_state;
+}
+
+// heartbeat thread function
+void jl_heartbeat_threadfun(void *arg)
+{
+    int s, ns = 1e9 - 1;
+    uint64_t t;
+
+    jl_adopt_thread();
+    jl_task_t *ct = jl_current_task;
+    jl_ptls_t ptls = ct->ptls;
+    uint8_t gc_state = jl_gc_safe_enter(ptls);
+
+    for (;;) {
```

Comment:
```suggestion
    while (true) {
```
IMO this is better style 🤷

Reply:
Some C compilers used to (still do?) complain about constants in loop conditions. But not about empty `for` statements. 🤷‍♂️

Reply:
haha gross. Okay, cool, feel free to ignore this then.